File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/sumEvenGrandparent_1315.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class sumEvenGrandparent_1315 {

    public int sumEvenGrandparent(TreeNode root) {
        return helper(root, false, false);
    }

    private int helper(TreeNode root, boolean parent, boolean grand) {
        if (root == null) return 0;
        int l = helper(root.left, root.val % 2 == 0, parent);
        int r = helper(root.right, root.val % 2 == 0, parent);
        if (grand) return l + r + root.val;
        else return l + r;
    }
}

File: Clion/ZeroTrac/Tree/sumEvenGrandparent_1315.cpp
//
// Created by TzashiNorpu on 10/25/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::sumEvenGrandparent(TreeNode *root) {
  int sum = 0;
  function<void(TreeNode *, int, int)> dfs = [&](TreeNode *node, int p, int pp) {
      if (!node) return;
      if (pp % 2 == 0) sum += node->val;
      dfs(node->left, node->val, p);
      dfs(node->right, node->val, p);
  };
  dfs(root, -1, -1);
  return sum;
}